### PR TITLE
Added functionality to reset failed attempts on successful login.

### PR DESCRIPTION
**Pull request by arjenzijlstra**
Adds functionality to reset the number of failed attempts on a successful login. The functionality can be enabled by setting `AXES_RESET_ON_SUCCESS` to true. The reset function makes use of `_query_user_attempts` to get the attempts to delete.

Also discussed in #352.

### CODE REVIEW

File: axes/attempts.py [hunk 1]
```diff
@@ -130,6 +130,13 @@ def get_user_attempts(request):
     return attempts
 
 
+def reset_user_attempts(request):
+    attempts = _query_user_attempts(request)
+    count, _ = attempts.delete()
+
```
Comment (markddavidoff) [inline axes/attempts.py]: also reset cached attempts. same issue as https://github.com/jazzband/django-axes/issues/363

Comment (arjenzijlstra) [reply axes/attempts.py]: I will add this today or tomorrow.

Comment (markddavidoff) [reply axes/attempts.py]: disregard, I guess we are already doing this.

Comment (arjenzijlstra) [reply axes/attempts.py]: I noticed, let me know if there is anything else to improve.

File: axes/models.py [hunk 2]
```diff
@@ -76,6 +76,7 @@ class Meta(object):
         verbose_name = _('access attempt')
         verbose_name_plural = _('access attempts')
 
+
```
Comment (aleksihakli) [inline axes/models.py]: Just stating that stylistic changes should not be mixed with functional changes; makes tracking regressions and bisecting harder.

File: axes/signals.py [hunk 3]
```diff
@@ -146,6 +147,13 @@ def log_user_logged_in(sender, request, user, **kwargs):  # pylint: disable=unus
             trusted=True,
         )
 
+    if settings.AXES_RESET_ON_SUCCESS:
+        count = reset_user_attempts(request)
+        log.info(
+            'AXES: Deleted %d failed login attempts by %s.', count,
```
Comment (aleksihakli) [inline axes/signals.py]: Preferably put the `count` argument on its own line.

File: axes/tests/test_access_attempt.py [hunk 4]
```diff
@@ -394,3 +394,49 @@ def test_custom_authentication_backend(self):
...
+        # But we should get one now
+        response = self._login()
+        self.assertContains(response, self.LOCKED_MESSAGE, status_code=403)
+
+    @override_settings(AXES_RESET_ON_SUCCESS=True)
+    def test_reset_on_success(self):
+        """Tests that the failure attempts resets after one successful
+        attempt when using the corresponding setting.
+        """
+        print('RESET = TRUE')
```
Comment (aleksihakli) [inline axes/tests/test_access_attempt.py]: Obsolete print statement.

File: axes/tests/test_access_attempt.py [hunk 5]
```diff
@@ -394,3 +394,49 @@ def test_custom_authentication_backend(self):
         request.user = self.user
         authenticate(request=request, foo='bar')
         self.assertEqual(AccessLog.objects.all().count(), 0)
+
+    # by default, AXES_RESET_ON_SUCCESS = False
+    def test_reset_on_success_default(self):
+        """Tests that the failure attempts does not reset after one successful
+        attempt by default.
+        """
+        print('RESET = FALSE')
```
Comment (aleksihakli) [inline axes/tests/test_access_attempt.py]: Obsolete print statement.

File: axes/tests/test_access_attempt.py [hunk 6]
```diff
@@ -394,3 +394,49 @@ def test_custom_authentication_backend(self):
...
+        # test until one try before the limit
+        for _ in range(1, settings.AXES_FAILURE_LIMIT):
```
Comment (aleksihakli) [inline axes/tests/test_access_attempt.py]: Just `range(settings.AXES_FAILURE_LIMIT - 1)` would be cleaner to read IMO, but that's just a preference.

File: axes/tests/test_access_attempt.py [hunk 7]
```diff
@@ -394,3 +394,49 @@ def test_custom_authentication_backend(self):
         request.user = self.user
         authenticate(request=request, foo='bar')
         self.assertEqual(AccessLog.objects.all().count(), 0)
+
+    # by default, AXES_RESET_ON_SUCCESS = False
+    def test_reset_on_success_default(self):
```
Comment (aleksihakli) [inline axes/tests/test_access_attempt.py]: This is exactly the same implementation as the function above this. Refactor these into a single common function that is called from both test cases and not prefixed `test` so it doesn't get executed automatically. Something like `_assert_resets_on_success` or similar could be neat.

File: axes/models.py (hunk 2 above)
Comment (arjenzijlstra) [reply axes/models.py]: I agree, shall I remove it from the PR?

File: axes/tests/test_access_attempt.py (hunk 6 above)
Comment (arjenzijlstra) [reply axes/tests/test_access_attempt.py]: I agree. I took it from another test and figured it was the style of this repos. I will rewrite it.

File: axes/tests/test_access_attempt.py (hunk 7 above)
Comment (arjenzijlstra) [reply axes/tests/test_access_attempt.py]: I refactored the tests into a common function, I hope you like the way I did it. Let me know if I can do something in a better way.

File: axes/models.py (hunk 2 above)
Comment (aleksihakli) [reply axes/models.py]: Either is OK, just making a note :)

File: axes/tests/test_access_attempt.py (hunk 7 above)
Comment (aleksihakli) [reply axes/tests/test_access_attempt.py]: Looks good to me.

**Comment by aleksihakli** [comment]
Looks good, merging into upstream.

**Pull request merged**